instrumentation/opentelemetry-instrumentation-asyncio/src/opentelemetry/instrumentation/asyncio/instrumentation_state.py

```diff
@@ -0,0 +1,60 @@
+"""
+Instrumentation State Tracker
+
+This module provides helper functions to safely track whether a coroutine,
+Future, or function has already been instrumented by the OpenTelemetry
+asyncio instrumentation layer.
+
+Some Python objects (like coroutines or functions) may not support setting
+custom attributes or weak references. To avoid memory leaks and runtime
+errors, this module uses a WeakKeyDictionary to safely track instrumented
+objects.
+
+If an object cannot be weak-referenced, it is silently skipped.
+
+Usage:
+    if not _is_instrumented(obj):
+        _mark_instrumented(obj)
+        # instrument the object...
+"""
+
+import weakref
+
+# A global WeakKeyDictionary to track instrumented objects.
+# Entries are automatically removed when the objects are garbage collected.
+_instrumented_tasks = weakref.WeakKeyDictionary()
+
+
+def _is_instrumented(obj) -> bool:
```

Comment:
If you can give a more specific type for `obj` please do. If it is unbounded or there are a tons of different types being instrumented, just add an explicit `Any`.

We are planning to make typing **opt-out** for any new files.